ods_ci/tests/Tests/100__deploy/130__operators/150__dsp_operator/150__dsp_operator.robot

```diff
@@ -0,0 +1,202 @@
+    ...    oc get datasciencecluster ${DSC_NAME}
+    IF  ${rc} == 0
+        Set Global Variable    ${DSC_EXISTS}    True
+        &{custom_cmp} =       Create Dictionary
+        ${rc}    ${out}=    Run And Return Rc And Output
+        ...    oc get datasciencecluster ${DSC_NAME} -o jsonpath='{.spec.components}'
+        ${custom_cmp}=    Load Json String    ${out}
+        Set Test Variable    ${custom_cmp}    ${custom_cmp}
+    ELSE
+        Set Global Variable    ${DSC_EXISTS}    False
+    END
+
+Fetch Image Url And Update Channel
+    [Documentation]    Fetch url for image and Update Channel
+    # Fetch subscription first
+    ${rc}    ${out}=    Run And Return Rc And Output
+    ...    oc get subscription ${OPERATOR_SUBSCRIPTION_NAME} -n ${OPERATOR_NAMESPACE} -o jsonpath='{.spec.source}'
+    Should Be Equal As Integers    ${rc}    0
+    Set Global Variable    ${CS_NAME}    ${out}
+    ${rc}    ${out}=    Run And Return Rc And Output
+    ...    oc get subscription ${OPERATOR_SUBSCRIPTION_NAME} -n ${OPERATOR_NAMESPACE} -o jsonpath='{.spec.sourceNamespace}'
+    Should Be Equal As Integers    ${rc}    0
+    Set Global Variable    ${CS_NAMESPACE}    ${out}
+    # Get CatalogSource
+    ${rc}    ${out}=    Run And Return Rc And Output
+    ...    oc get catalogsource ${CS_NAME} --namespace ${CS_NAMESPACE} -o jsonpath='{.spec.image}'
+    Should Be Equal As Integers    ${rc}    0
+    Set Global Variable    ${IMAGE_URL}    ${out}
+    ${rc}    ${out}=    Run And Return Rc And Output
+    ...    oc get subscription ${OPERATOR_SUBSCRIPTION_NAME} --namespace ${OPERATOR_NS} -o jsonpath='{.spec.channel}'
+    Should Be Equal As Integers    ${rc}    0
+    Set Global Variable    ${UPDATE_CHANNEL}    ${out}
+
+Operator Deployment Should Be Ready
+    [Documentation]    Loop until the Operator deployment is ready
+    ${rc}=    Set Variable    1
+    TRY
+        WHILE    ${rc} != 0    limit=10m
+            Sleep    5s
+            ${rc}    ${output}=    Run And Return Rc And Output
+            ...    oc wait --for condition=available -n ${OPERATOR_NAMESPACE} deploy/${OPERATOR_DEPLOYMENT_NAME}
+        END
+    EXCEPT    WHILE loop was aborted    type=start
+        Fail    msg=Operator deployment did not get ready
+    END
+
+Restore Datasciencecluster If Existed
+    [Documentation]     Restore Datasciencecluster If Existed as it was in case it existed previously
+    IF      ${DSC_EXISTS} == True
+        Wait For Datasciencecluster Crd To Be Available
+        Wait For Mutatingwebhook To Be Available
+        Apply DataScienceCluster CustomResource     ${DSC_NAME}     True    ${custom_cmp}
+    END
+
+Cleanup Olm Install Dir
+    [Documentation]    Clean up olm install repo
+    ${return_code}=    Run and Watch Command    rm -vRf ${EXECDIR}/${OLM_DIR}    timeout=10 min
+    Should Be Equal As Integers    ${return_code}    0    msg=Error while cleaning up olminstall directory
+
+Teardown Detect Pre-existing Install Of Argo Workflows
```

Comment:
if the test fails before re-installing RHOAI, the cluster won't have RHOAI operator installed hence this teardown and the rest of the tests will fail. Could you have RHOAI re-installed if broken in the teardown as well>

Reply:
the problem is that teardown is always executed no matter whether the test case fails or is successful, and if it is successful the operator is already installed. The keywords that are prone to fail come after reinstalling RHOAI:
```
Detect Pre-existing Install Of Argo Workflows And Block RHOAI Install
    [Documentation]    Detect Pre-existing Install Of Argo Workflows And Block RHOAI Install
    [Tags]                  Operator                ODS-2651                Tier3
    Fetch Image Url And Update Channel
    Check Whether DSC Exists And Save Component Statuses
    Fetch Cluster Type By Domain
    IF    "${CLUSTER_TYPE}" == "selfmanaged"
        Uninstall RHODS In Self Managed Cluster
        Create Argo Workflow From Template
        Install RHODS In Self Managed Cluster Using CLI     ${CLUSTER_TYPE}     ${IMAGE_URL}

    ELSE IF    "${CLUSTER_TYPE}" == "managed"
        Uninstall RHODS In OSD
        Create Argo Workflow From Template
        Install RHODS In Managed Cluster Using CLI      ${CLUSTER_TYPE}     ${IMAGE_URL}
    END
    Restore Datasciencecluster If Existed
    Wait For Failed Conditions
    ${return_code}          ${output}               Run And Return Rc And Output
    ...                     oc delete DataScienceCluster default-dsc
    Log To Console          ${output}
    Should Be Equal As Integers
    ...                     ${return_code}
    ...                     0
    ...                     msg=Error deleting DataScienceCluster CR
    [Teardown]      Teardown Detect Pre-existing Install Of Argo Workflows
```

Reply:
cannot you check if RHOAI is installed in the teardown to decide if to install or not?